compiler/circle-mpqsolver/src/bisection/Evaluator.cpp

```diff
@@ -0,0 +1,134 @@
+  importer.importGroup("value");
+
+  bool is_raw_data = importer.isRawData();
+
+  const auto num_records = importer.numData();
+  if (num_records == 0)
+    throw std::runtime_error("The input data file does not contain any record.");
+  const auto input_nodes = loco::input_nodes(module->graph());
+  const auto num_inputs = input_nodes.size();
+
+  WholeOutput dataset_output;
+
+  // Create interpreter.
+  luci_interpreter::Interpreter interpreter(module);
+  for (int32_t record_idx = 0; record_idx < num_records; record_idx++)
+  {
+    if (num_inputs != importer.numInputs(record_idx))
+      throw std::runtime_error("Wrong number of inputs.");
+    for (int32_t input_idx = 0; input_idx < num_inputs; input_idx++)
+    {
+      const auto *input_node = loco::must_cast<const luci::CircleInput *>(input_nodes[input_idx]);
+      assert(input_node->index() == input_idx);
+
+      std::vector<char> input_data(get_tensor_size(input_node));
+
+      if (!is_raw_data)
+      {
+        loco::DataType dtype;
+        Shape shape;
+        importer.readTensor(record_idx, input_idx, &dtype, &shape, input_data.data());
+      }
+      else
+      {
+        // Skip type/shape check for raw data
+        importer.readTensor(record_idx, input_idx, input_data.data());
+      }
+
+      interpreter.writeInputTensor(input_node, input_data.data(), input_data.size());
+    }
+
+    interpreter.interpret();
+
+    Output nn_output;
+
+    // Get output.
+    const auto output_nodes = loco::output_nodes(module->graph());
+    for (int i = 0; i < module->graph()->outputs()->size(); i++)
+    {
+      const auto *output_node = loco::must_cast<const luci::CircleOutput *>(output_nodes[i]);
+      Buffer output_data(get_tensor_size(output_node));
+      interpreter.readOutputTensor(output_node, output_data.data(), output_data.size());
+      // output
+      nn_output.push_back(output_data);
+    }
+    dataset_output.push_back(nn_output);
+  }
+
+  return dataset_output;
+}
+
+} // namespace
+
+DatasetEvaluator::DatasetEvaluator(const luci::Module *ref_module, const std::string &h5file,
+                                   const ErrorMetric &metric)
+  : _ref_module(ref_module), _h5file(h5file), _metric(&metric)
+{
+  _ref_output = compute_outputs(_ref_module, _h5file);
+}
+
+void DatasetEvaluator::validate(const luci::Module *trgt_fq_module) const
+{
+  const auto output_nodes = loco::output_nodes(trgt_fq_module->graph());
+  for (size_t out_index = 0; out_index < output_nodes.size(); ++out_index)
+  {
+    const auto *output_node = loco::must_cast<const luci::CircleOutput *>(output_nodes[out_index]);
+    loco::DataType out_dtype = output_node->dtype();
+    if (out_dtype != loco::DataType::FLOAT32)
+      throw std::runtime_error("Unsupported output dtype. " + output_node->name());
+  }
+}
+
+float DatasetEvaluator::evaluate(const luci::Module *trgt_fq_module) const
+{
+  if (_metric == nullptr)
+    throw std::runtime_error("Unsupported metric. ");
```

Comment:
Please remove empty space at the end.
- maybe `Invalid metric` ?

optional) null check for `trgt_fq_module`

Reply:
Ahh... sorry. ok.

Reply:
I think `_metric` validation should be done when set, not when used.

Reply:
fixed